File: rusty/scaly/src/containers/array.rs
extern crate libc;

use self::libc::c_void;
use self::libc::memcpy;
use containers::{Ref, Vector};
use memory::Page;
use std::mem::{align_of, size_of};
use std::ops::{Deref, DerefMut, Index};
use std::ptr::{null_mut, write};

#[derive(Copy, Clone)]
pub struct Array<T: Copy> {
    vector: Vector<T>,
    length: usize,
}

impl<T: Copy> Array<T> {
    pub fn new() -> Array<T> {
        Array {
            vector: Vector {
                data: null_mut(),
                length: 0,
            },
            length: 0,
        }
    }

    pub fn get_buffer(&self) -> *const T {
        self.vector.data
    }

    pub fn get_length(&self) -> usize {
        self.length
    }

    pub fn get_capacity(&self) -> usize {
        self.vector.length
    }

    pub fn add(&mut self, item: T) {
        if self.length == self.vector.length {
            self.reallocate();
        }

        unsafe {
            let location = self.vector.data.offset(self.length as isize);
            write(location, item);
        }

        self.length += 1;
    }

    pub fn reallocate(&mut self) {
        let _own_page = Page::own(Ref::from(self as *mut Array<T>));
        let size = size_of::<T>();
        unsafe {
            if self.vector.length == 0 {
                let exclusive_page = (*_own_page).allocate_exclusive_page();
                let capacity = (*exclusive_page).get_capacity::<T>();
                self.vector.length = capacity / size;
                self.vector.data = (*exclusive_page)
                    .allocate_raw(self.vector.length * size, align_of::<T>())
                    as *mut T;
            } else {
                let old_length = self.vector.length;
                self.vector.length *= 2;
                let old_data = self.vector.data;
                self.vector.data =
                    (*_own_page).allocate_raw(self.vector.length * size, align_of::<T>()) as *mut T;
                memcpy(
                    self.vector.data as *mut c_void,
                    old_data as *mut c_void,
                    old_length * size,
                );
                let old_exclusive_page = Page::get(old_data as usize);
                (*old_exclusive_page).forget();
            }
        }
    }
}

impl<T: Copy> Index<usize> for Array<T> {
    type Output = T;
    fn index(&self, offset: usize) -> &Self::Output {
        unsafe { &*self.vector.data.offset(offset as isize) }
    }
}

impl<T: Copy> Deref for Array<T> {
    type Target = [T];
    fn deref(&self) -> &[T] {
        unsafe { ::std::slice::from_raw_parts(self.vector.data, self.length) }
    }
}

impl<T: Copy> DerefMut for Array<T> {
    fn deref_mut(&mut self) -> &mut [T] {
        unsafe { ::std::slice::from_raw_parts_mut(self.vector.data, self.length) }
    }
}

File: runtime/containers/Array.cpp
namespace scaly {
namespace containers {

using namespace scaly::memory;

template<class T> struct Array : Object {
    size_t length;
    Vector<T>* vector;

    Array() : length(0), vector(nullptr) {}

    T* get_buffer() {
        if (this->vector == nullptr)
            return nullptr;
        return this->vector->data;
    }

    size_t get_length() {
        return this->length;
    }

    size_t get_capacity() {
        if (this->vector ==  nullptr)
            return 0;
        return this->vector->length;
    }

    void add(T item) {
        if (this->vector == nullptr || this->length == this->vector->length) {
            reallocate();
        }

        auto location = this->vector->data + this->length;
        *location = item;

        this->length++;
    }

    void add(const Vector<T>& items) {
        auto length = this->length + items.length;
        auto size = sizeof(T);
        if (this->vector == nullptr || length >= this->vector->length) {
            reallocate();
        }
        if (length >= this->vector->length) {
            auto own_page = Page::get(this);
            auto data = (T*)own_page->allocate_raw(length * size, alignof(T));
            auto bytes_to_copy = this->vector->length * size;
            if (this->vector->length > 0)
                memcpy(data, this->vector->data, bytes_to_copy);
            Page* old_exclusive_page = Page::get(this->vector->data);
            own_page->deallocate_exclusive_page(old_exclusive_page);
            this->vector->data = data;
            this->vector->length = length;
        }

        if (items.length > 0)
            memcpy(this->vector->data + this->length * size, items.data, items.length * size);
        this->length += items.length;
    }

    T* get(size_t i) {
        if (i >= this->length)
            return nullptr;

        return (*this)[i];
    }

    T* operator [](size_t i) {
        if (this->length < i + 1)
            return nullptr;
        return this->vector->data + i;
    }

    void reallocate() {
        auto own_page = Page::get(this);
        auto size = sizeof(T);
        size_t length;
        if (this->vector == nullptr) {
            Page* exclusive_page = own_page->allocate_exclusive_page();
            auto capacity = exclusive_page->get_capacity(alignof(T)) - sizeof(Vector<T>);
            length = capacity / size;
            this->vector = new(alignof(Vector<T>), own_page) Vector<T>((T*)nullptr, length);
            this->vector->data = (T*)exclusive_page->allocate_raw(capacity, alignof(T));             
        } else {
            length = this->vector->length * 2;
            auto data = (T*)own_page->allocate_raw(length * size, alignof(T));
            auto bytes_to_copy = this->vector->length * size;
            memcpy(data, this->vector->data, bytes_to_copy);
            Page* old_exclusive_page = Page::get(this->vector->data);
            own_page->deallocate_exclusive_page(old_exclusive_page);
            this->vector->data = data;
            this->vector->length = length;
        }
    }
};

}
}